src/Core.jl

```diff
@@ -56,30 +56,82 @@ offset(p::Node, xy) = chxy(p, getxy(p) .+ xy)
 const WeightedNode{T<:Real} = Node{T}
 const UnWeightedNode = Node{ONE}
 
+############################ Grid Types ############################
+# Abstract grid geometry types
+abstract type AbstractGridType end
+
+struct SquareGrid <: AbstractGridType end
+struct TriangularGrid <: AbstractGridType end
+
 ############################ GridGraph ############################
-# GridGraph
-struct GridGraph{NT<:Node}
+# Main definition
+struct GridGraph{NT<:Node, GT<:AbstractGridType}
+    gridtype::GT
     size::Tuple{Int,Int}
     nodes::Vector{NT}
     radius::Float64
 end
+
+is_square_grid(g::GridGraph) = g.gridtype isa SquareGrid
+is_triangular_grid(g::GridGraph) = g.gridtype isa TriangularGrid
+
+# Base constructors (for any node/grid type)
+GridGraph(size::Tuple{Int,Int}, nodes::Vector{NT}, radius::Real) where {NT<:Node} =
+    GridGraph(SquareGrid(), size, nodes, radius)
+
+GridGraph(gridtype::GT, size::Tuple{Int,Int}, nodes::Vector{NT}, radius::Real) where {NT<:Node, GT<:AbstractGridType} =
+    GridGraph{NT, GT}(gridtype, size, nodes, radius)
+
 function Base.show(io::IO, grid::GridGraph)
-    println(io, "$(typeof(grid)) (radius = $(grid.radius))")
+    gridtype_name = grid.gridtype isa SquareGrid ? "Square" : "Triangular"
+    println(io, "$(gridtype_name)$(typeof(grid)) (radius = $(grid.radius))")
     print_grid(io, grid; show_weight=SHOW_WEIGHT[])
 end
 Base.size(gg::GridGraph) = gg.size
 Base.size(gg::GridGraph, i::Int) = gg.size[i]
 function graph_and_weights(grid::GridGraph)
-    return unit_disk_graph(getfield.(grid.nodes, :loc), grid.radius), getfield.(grid.nodes, :weight)
+    if is_triangular_grid(grid)
+        # For triangular grids, use physical positions
+        physical_locs = [physical_position(node) for node in grid.nodes]
```

Comment:
You do not need to use dispatch, just implement the same `physical_position` interface for square lattice.